Ckan upgrade 2.8.0a - permit alternative DistinguishedName (dn)

lucasbenevides opened the pull request:
Includes a new parameter, called **ckanext.ldap.base_dn_alt** with permits a new Distinguished Name to be searched when a login operation is performed. 
It uses the same logic implemented in the parameter ckanext.ldap.search_alt that already existed.
This was implemented because the forest of my organization has two organizational units (OU) and the CKAN must be acessed from users from both OUs.

LucasBenevidev committed:
Add the parameter ckanet.ldap.base_dn_alt.

Adds to the description the alternative base Distinguished Name (DN) in case the forest has two Organizational Units (OU) to look for. The way the python-ldap package is built, it needs the DN to have an OU. The original ldap library in C does not need it. Although, as a corner case solution, a new parameter of the alternative DN was created.

LucasBenevidev committed:
Include the parameter ckanext.ldap.base_dn_alt.

Include the alternative Distinguished Name (dn) parameter. It seems (needs confirmation) like the way the python-ldap package was built, it does not accept a DN wit only the Domain Component (DC), it must have a OU. So the parameter ckanext.ldap.base_dn_alt will be useful in case the Active Directory (AD) forest has two Organizational Units. Although, the best solution would be changing the python-ldap package so that the search functions worked with a DN without an OU.

LucasBenevidev committed:
Added new parameter ckanext.ldap.base_dn_alt

This modification permits the use of two base Distinguished Names that will be searched for. It works along with the ckanext.ldap.search_alt, using the same logic. It there was already a way to use two Organizational Units in the cnakext.ldap.base_dn this change would be dull.

jrdh commented on ckanext/ldap/controllers/user.py:
```diff
@@ -321,14 +329,15 @@ def _ldap_search(cnx, filter_str, attributes, non_unique=u'raise'):
+        res = cnx.search_s(base_dn_str, ldap.SCOPE_SUBTREE,
                            filterstr=filter_str, attrlist=attributes)
+        '''res = cnx.search_s(None, ldap.SCOPE_SUBTREE, filterstr=filter_str, attrlist=attributes)
+        '''
```
can the commented out code string be removed please?

lucasbenevides replied:
Sure it can. Should not have been committed.

jrdh commented on ckanext/ldap/controllers/user.py:
```diff
@@ -292,18 +292,26 @@ def _find_ldap_user(login):
+                login=ldap.filter.escape_filter_chars(login))
+                ret = _ldap_search(cnx, filter_str, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'raise')
+        elif ret is None and u'ckanext.ldap.base_dn_alt' in config:
+            ret = _ldap_search(cnx, filter_str, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'log')
```
If I'm reading this right, the logic here doesn't process all possible combinations of base DN and search. We try:

- `base_dn` + `search` on line 297
- `base_dn` + `search.alt`: on line 301
- `base_dn_alt` + `search` on line 305 and 307.

but not `base_dn_alt` + `search.alt`.

lucasbenevides replied:
Yes, you are correct. I have changed this code. I created a variable calles filter_str_alt in order to make the code more clear. See if you like it.
Will change my repository and should I make a new pull request?

jrdh replied:
Looks much neater, nice work!

jrdh commented on ckanext/ldap/controllers/user.py:
```diff
@@ -321,14 +329,15 @@ def _ldap_search(cnx, filter_str, attributes, non_unique=u'raise'):
     except ldap.OPERATIONS_ERROR as e:
-        log.error(
-            u'LDAP query failed. Maybe you need auth credentials for performing searches? Error returned by the server: ' + e.info)
+        log.error(u'LDAP query failed. Maybe you need auth credentials for performing searches?')
```
Can the error info be added back in? Or another way or getting more information about the error for the logs?

lucasbenevides replied:
The e.info does not exist. I was getting a Notice in the error log, that says the info atribute is not available when it did enter in that except. 
I don't know how I could get more information to de logs.

```

jrdh replied:
The ldap library's doc isn't all that clear but I think from my quick look the `info` attribute is optional and dependant on whether the server returned an extra info or not. Perhaps we could check if there is an `info` attribute available or not and then log it or not based on that?

jrdh reviewed (changes requested):
Just a couple of minor queries. Apologies for the delay getting this reviewed!

LucasBenevidev committed:
Correct when dn_alt and search_str_alt 

Removes commented like as indicated by @jrdh.
Corrects the case when alternative filter search is changed and distinguished name alternative must be tested.

lucasbenevides commented:
I committed to my respository branch. I think it automatically brings the changes.

jrdh commented on ckanext/ldap/controllers/user.py:
```diff
@@ -287,6 +287,8 @@ def _find_ldap_user(login):
     filter_str = config[u'ckanext.ldap.search.filter'].format(
         login=ldap.filter.escape_filter_chars(login))
+    filter_str_alt = config[u'ckanext.ldap.search.alt'].format(
+        login=ldap.filter.escape_filter_chars(login))
```
`ckanext.ldap.search.alt` is optional so I'm pretty sure this will raise an exception if it's not defined in the config.

alycejenni commented on ckanext/ldap/controllers/user.py:
```diff
@@ -106,7 +106,7 @@ def _login_success(self, user_name, came_from):
         session[u'ckanext-ldap-user'] = user_name
         session.save()
-        toolkit.redirect_to(u'user.logged_in', came_from=came_from)
+        toolkit.redirect_to(controller=u'user', action=u'logged_in', came_from=came_from)
```
This reverts back to the old pylons syntax.
```suggestion
        toolkit.redirect_to(u'user.logged_in', came_from=came_from)
```

alycejenni commented on ckanext/ldap/controllers/user.py:
```diff
@@ -321,14 +327,14 @@ def _ldap_search(cnx, filter_str, attributes, non_unique=u'raise'):
     except ldap.OPERATIONS_ERROR as e:
-        log.error(
-            u'LDAP query failed. Maybe you need auth credentials for performing searches? Error returned by the server: ' + e.info)
+        log.error(u'LDAP query failed. Maybe you need auth credentials for performing searches?')
```
```suggestion
		error_msg = u'LDAP query failed. Maybe you need auth credentials for performing searches?'
		if hasattr(e, u'info'):
			error_msg += u' Error returned by the server: ' + e.info
        log.error(error_msg)
```

alycejenni commented on ckanext/ldap/controllers/user.py:
```diff
@@ -287,23 +287,29 @@ def _find_ldap_user(login):
-            ret = _ldap_search(cnx, filter_str, attributes, non_unique=u'raise')
+            ret = _ldap_search(cnx, filter_str_alt, attributes, config[u'ckanext.ldap.base_dn'], non_unique=u'raise')
+        if ret is None and u'ckanext.ldap.base_dn_alt' in config:
+            ret = _ldap_search(cnx, filter_str, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'log
```
```suggestion
            ret = _ldap_search(cnx, filter_str, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'log')
```

alycejenni commented on ckanext/ldap/controllers/user.py:
```diff
@@ -287,23 +287,29 @@ def _find_ldap_user(login):
+        if ret is None and u'ckanext.ldap.base_dn_alt' in config:
+            ret = _ldap_search(cnx, filter_str, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'log
+        if ret is None and u'ckanext.ldap.base_dn_alt' in config and u'ckanext.ldap.search.alt' in config:
+            ret = _ldap_search(cnx, filter_str_alt, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'raise
```
```suggestion
            ret = _ldap_search(cnx, filter_str_alt, attributes, config[u'ckanext.ldap.base_dn_alt'], non_unique=u'raise')
```

alycejenni commented on ckanext/ldap/controllers/user.py:
```diff
@@ -287,23 +287,29 @@ def _find_ldap_user(login):
 
     filter_str = config[u'ckanext.ldap.search.filter'].format(
         login=ldap.filter.escape_filter_chars(login))
+    filter_str_alt = config[u'ckanext.ldap.search.alt'].format(
```
```suggestion
    filter_str_alt = config.get(u'ckanext.ldap.search.alt', u'').format(
```
